src/main/java/org/breedinginsight/services/parsers/excel/ExcelParser.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.breedinginsight.services.parsers.excel;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.poi.ss.usermodel.Cell;
+import org.apache.poi.ss.usermodel.CellType;
+import org.apache.poi.ss.usermodel.Row;
+import org.apache.poi.ss.usermodel.Sheet;
+import org.breedinginsight.services.parsers.ParsingException;
+
+import java.util.*;
+
+/*
+ * This xls/xlsx parser expects spreadsheets with the first row as column names and subsequent rows
+ * as data. Only data in a column with a column name will be kept, data outside these columns is
+ * discarded.
+ */
+@Slf4j
+public class ExcelParser {
+
+    private static final int EXCEL_COLUMN_NAMES_ROW = 0;
+
+    public List<ExcelRecord> parse(Sheet sheet, Set<String> columns) throws ParsingException {
+
+        List<ExcelRecord> records = new ArrayList<>();
+
+        Row columnNames = sheet.getRow(EXCEL_COLUMN_NAMES_ROW);
+
+        if (columnNames == null) {
+            throw new ParsingException("Missing column names row");
+        }
+
+        Map<Integer, String> indexColNameMap = new HashMap<>();
+
+        // get column name to index mapping
+        for(int colIndex=0; colIndex<columnNames.getLastCellNum(); colIndex++) {
+            Cell cell = columnNames.getCell(colIndex, Row.MissingCellPolicy.RETURN_BLANK_AS_NULL);
+            if (cell.getCellType() != CellType.STRING) {
+                throw new ParsingException("Column name must be string cell");
+            }
+            if (cell != null) {
+                indexColNameMap.put(colIndex, cell.getStringCellValue());
+            } else
+            {
+                // not throwing an exception here in case they have empty cells to right of columns or even
+                // between as we're not enforcing that, just that all the expected columns are present
+                log.info("Ignoring blank header column");
+            }
+        }
+
+        // check for duplicates
+        if (hasDuplicates(new ArrayList(indexColNameMap.values()))) {
+            throw new ParsingException("Found duplicate column names");
+        }
```

Comment:
If it doesn't already exist, seems like a good unit test to add as well

Reply:
I'll check it out and add a test case

Reply:
Ya, this is because I allowed blank columns to accommodate for leaving accidental empty columns and those are included in the duplicate check although they shouldn't be. I think I just won't allow empty columns as that's the simplest way to handle this.

Reply:
Actually looking at this more, the getCell call wasn't working as I thought so I will still allow empty header cells, they will just not be included in the column headers and will therefore not be checked for duplicates.

Reply:
Fixed, added tests, and pushed changes.